CloundFoundry connector

xtremerui opened the pull request:
Support for authenticating against [CloudFoundry](https://www.cloudfoundry.org/)

Documentation of [CloudFoundry UAA server](https://docs.cloudfoundry.org/api/uaa/version/74.4.0/index.html#overview)

cc @pivotal-jwinters

Commit:
Add new connector for Cloudfoundry

- Verifies user is part of orgs and spaces for group claims

Signed-off-by: Joshua Winters <jwinters@pivotal.io>
Co-authored-by: Shash Reddy <sreddy@pivotal.io>

Commit:
update cf connector to use 'authorization_endpoint' from /v2/info

Co-authored-by: Topher Bullock <tbullock@pivotal.io>
Signed-off-by: Josh Winters <jwinters@pivotal.io>

daniellavoie committed:
Added support for CF resources pagination

Signed-off-by: Daniel Lavoie <dlavoie@live.ca>

Commit:
add unit test and api call to `audited_spaces` and `managed_spaces`

Signed-off-by: Zoe Tian <ztian@pivotal.io>
Co-authored-by: Ciro S. Costa <cscosta@pivotal.io>
Signed-off-by: w3tian <w3tian@uwaterloo.ca>

Commit:
append role to space guids

Signed-off-by: Rui Yang <ryang@pivotal.io>
Co-authored-by: Joshua Winters <jwinters@pivotal.io>

Commit:
cf: add org to groups claims

Signed-off-by: Joshua Winters <jwinters@pivotal.io>
Co-authored-by: Rui Yang <ryang@pivotal.io>

Commit:
cf: add org guid to groups claims

Co-authored-by: Rui Yang <ryang@pivotal.io>
Signed-off-by: Joshua Winters <jwinters@pivotal.io>

Commit:
add cf org:space:role group claim to token

Signed-off-by: Joshua Winters <jwinters@pivotal.io>
Co-authored-by: Rui Yang <ryang@pivotal.io>

Commit:
fix lint errors

gofumpt-ed

Signed-off-by: Rui Yang <ryang@pivotal.io>

xtremerui commented:
Hi, just checking back. Since there is no feedback here yet so I am wondering are there certain requirements to meet for Dex to add a new connector? Is there a way to move forward for this PR or its just a simple `Nah not this connector`? 

Thank you for your time.

Albertoimpl commented:
Our team is also interested in this connector to adopt dex.
I saw it used to be supported but seems to be gone now: https://github.com/dexidp/dex/pull/542/files
Is there anything we can do to move this forward? 
Thanks!

mvdkleijn commented:
@xtremerui This also has merge conflicts... :smile:

Commit:
run golangcli-lint

Signed-off-by: Rui Yang <ruiya@vmware.com>

xtremerui commented:
@sagikazarmark I kown there must be lots of  pending PRs in your list. Just wanted to get some feedback here so we could move forward. Thx!

Commit:
fix sanity check errors

Signed-off-by: Rui Yang <ruiya@vmware.com>

sagikazarmark commented:
Generally, we are against adding new connectors to the codebase (at least until #1635 is figured out), but I'd hate to see all this work going to waste.

Another reason that keeps us from adding all potential connectors is the lack of resources to support/test them.

If we were to add CloudFoundry as a connector, it would be labeled as a "community maintained" connector, until we can figure out a way to move it out of the core as a plugin of some sort. We would refer to you whenever someone reports an issue.

Would that be acceptable for you @xtremerui ?

xtremerui commented:
@sagikazarmark that's fair enough. Concourse team would have a better set of tooling to maintain this connector. Once connectors are out of the core, Concourse org would take ownership of cloudfoundry connector as well.

sagikazarmark commented on connector/cf/cf.go:
```diff
@@ -0,0 +1,394 @@
+package cf
```
I'd prefer if this would be something like `cloudfoundry`. Cf is too short and not descriptive.

sagikazarmark commented on connector/cf/cf.go:
```diff
@@ -0,0 +1,394 @@
+	InsecureSkipVerify bool     `json:"insecureSkipVerify"`
+}
+
+type CCResponse struct {
+	NextURL      string     `json:"next_url"`
+	Resources    []Resource `json:"resources"`
+	TotalResults int        `json:"total_results"`
+}
+
+type Resource struct {
+	Metadata Metadata `json:"metadata"`
+	Entity   Entity   `json:"entity"`
+}
+
+type Metadata struct {
+	GUID string `json:"guid"`
+}
+
+type Entity struct {
+	Name             string `json:"name"`
+	OrganizationGUID string `json:"organization_guid"`
+}
+
+type Space struct {
+	Name    string
+	GUID    string
+	OrgGUID string
+	Role    string
+}
+
+type Org struct {
+	Name string
+	GUID string
+}
```
Are these types exported for a reason?

xtremerui replied:
Not sure the reason behind it. I will update them.

sagikazarmark reviewed (changes requested):
A few nits

sagikazarmark commented:
@xtremerui let me know if you think this is ready.

And you are right: some sort of documentation would be nice. At least adding it to the list of connectors.

Commit:
rename connector;make types private;

Signed-off-by: Rui Yang <ruiya@vmware.com>

Commit:
add cloudfoundry to connector list in readme

Signed-off-by: Rui Yang <ruiya@vmware.com>

xtremerui commented:
@sagikazarmark the PR is updated. Thx for the reviewing.

We will submit a PR to dex docs too. BTW, in the readme `https://dexidp.io/docs/connectors/oauth/` is somehow broken  as in the connector listing, though the PR for oauth doc is already merged.

Kump3r committed:
Switching to CloudFoundry v3 API

Signed-off-by: Kump3r <tonevkalin@gmail.com>

Kump3r committed:
Add missing endpoint

Signed-off-by: Kump3r <tonevkalin@gmail.com>

Kump3r committed:
Refactor CloudFoundry API request handling to use updated response structure

Signed-off-by: Kump3r <tonevkalin@gmail.com>

Kump3r committed:
Use simple structures for the apiResult

Signed-off-by: Kump3r <tonevkalin@gmail.com>

IvanChalukov committed:
Switching to CF API V3 endpoints

Signed-off-by: IvanChalukov <ichalukov@gmail.com>

Kump3r committed:
Addapt cloudfoundry_test.go acceptance tests for CF API v3

Signed-off-by: Kump3r <tonevkalin@gmail.com>

taylorsilva committed:
update cf connector to use slog instead

Signed-off-by: Taylor Silva <dev@taydev.net>

backkem committed:
fix(mysql): quote `groups` reserved word in query replacer (#4580)

`groups` is a reserved word in MySQL >= 8.0.2, causing migration 13
to fail with a syntax error on `ALTER TABLE password ADD COLUMN groups`.

Fixes #4579

Signed-off-by: Michiel De Backker <mail@backkem.me>

matzegebbe committed:
fix(connector): update authproxy and oauth to match CallbackConnector interface (#4589)

The PKCE support added in v2.45.0 changed the CallbackConnector interface
signatures but missed updating the authproxy and oauth connectors. This
caused a type assertion failure in handleConnectorLogin(), resulting in
"Requested resource does not exist" errors when using these connectors.

Update LoginURL to return (string, []byte, error) and HandleCallback to
accept a []byte connData parameter for both connectors and their tests.

Signed-off-by: Mathias Gebbe <mathias.gebbe@gmail.com>

nabokihms committed:
Merge pull request #4604 from deckhouse/2.45.1-prepare-release

v2.45.1 prepare release

IvanChalukov committed:
Update CF connector to implement new Connector interface

Signed-off-by: IvanChalukov <ichalukov@gmail.com>

IvanChalukov committed:
fix status check on uaaResp

Signed-off-by: IvanChalukov <ichalukov@gmail.com>
